Draft for PR description

mastricker opened the pull request:
This is with in reference to #188 to put a description for our review process somewhere.

Please have a look at it, any changes/additions welcome.
I tried to be as concise as possible without imposing too many thing, because there will be noise.

One thing I tried to make clear is:
- If you start a review, see it through, you are responsible
- Single comments are ok, but should stay single comments

- Judge wisely on changes which happen after a PR, depending on the change.

_Edit_: just to be clear, this is not a Draft PR, is is a draft for the document for review :)

Luthaf commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,70 @@
+person.
+
+So once you have added something to the code base and checked locally if the
+checks pass merged and/or rebase on top of the current state of the master
```
I am not sure what this means. Is it "one should merge/rebase on top of master and then check that the tests pass?"

mastricker replied:
I reformulated it, hope this is clearer now.

Luthaf replied:
Yep, thanks!

max-veit commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,76 @@
+Further, we provide a PR template on github with several checkmarks to help
+start and streamline the process. Please make use of it and get rid of all
+unnecessary points, which just globber the PR message. To make it easy for
+possible reviewers do not hesitate to contact any of the people invovled in the
```
Not sure I like the wording of these last two sentences. First, we want to emphasize that the PR checklist is the _main_ thing you can do to make it easier for potential reviewers. Second, we want to say that you should also feel free to contact the librascal developers about a PR if you're unsure about something.

But I think we should rather encourage PRs as a way of _starting_ the conversation about a new feature or change, and encourage draft PRs (as I wrote in the contribution guidelines!) rather than just opening a branch and letting it rot.

Finally, it's a bit unfair to mention things being "easier to talk about in person" when we might soon have collaborators in other parts of the world who have no possibility of talking to us in person (meaning: face to face). It's fine to say "don't hesitate to contact us", but I'd leave out the bit about conversations in person, since we all pretty much know that already.

max-veit replied:
Oh, and "involved" (typo) if you decide to keep it

max-veit commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,76 @@
+development of librascal also beforehand. Some thing are easier to talk about in
+person.
+
+So once you have added something to the code base and checked locally if the
```
Remove "So". Also: sentence fragment; change "checked" to "please check" (or better: "please run all the checks locally and ensure they pass")

max-veit commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,76 @@
+adding new documentation, in short: contributing to the codebase in any form,
+the tests we have employed should all pass. These include tests for
+functionality on the C++ as well as the python side but also include checks for
+code style as far as they can be automated (see our :ref:`Conding convention
```
"Coding convention" (typo)

max-veit commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,76 @@
+person.
+
+So once you have added something to the code base and checked locally if the
+checks pass. Checks include the code tests as well as the checks for coding
```
"These checks include". Also, "checks" gets used a lot (with slightly different meanings), maybe keep "tests" as above

max-veit commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,76 @@
+
+So once you have added something to the code base and checked locally if the
+checks pass. Checks include the code tests as well as the checks for coding
+style as far as they can be checked automatically. You should also have added
```
Keep present tense, as it's clearer and more direct: "You should also add"

max-veit commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,76 @@
+So once you have added something to the code base and checked locally if the
+checks pass. Checks include the code tests as well as the checks for coding
+style as far as they can be checked automatically. You should also have added
+tests for your added functionality or adjusted tests, which conflict with a new
```
"functionality. Likewise, if your changes break existing tests, be sure to update (not disable!) the affected tests so that they pass."

max-veit commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,76 @@
+checks pass. Checks include the code tests as well as the checks for coding
+style as far as they can be checked automatically. You should also have added
+tests for your added functionality or adjusted tests, which conflict with a new
+design. The next step is to merge or rebase with the master branch. If
```
Just "Next, merge with the master branch." We don't want to be encouraging rebases unless it turns out it's necessary; also, a single final rebase will remove any extraneous merge commits.

Luthaf replied:
I disagree here: one should rebase before submitting a PR, to create it with a nice history. Once the PR is created, merges are fine since Github interface plays poorly with force pushes.

mastricker replied:
I tend to agree with Luthaf. Added a bracket-note to contact us if unsure which way.

max-veit commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,76 @@
+design. The next step is to merge or rebase with the master branch. If
+necessary, add changes to the code to comply with any other added tests coming
+in from the updated master branch if they were not resolved in the merging or
+rebase process. Then you are ready to create your PR. Upon creating the PR our
```
Here I disagree with the procedure itself: We should encourage creating PRs _as soon as the feature is ready for discussion_, not once all the technical checks have passed. If you want to put some changes up for discussion but don't have all the technical details sorted yet, then it's fine to make a draft PR (in fact, we should encourage it). This has the further advantage that the CI is run sooner, catching errors that the contributor might not have seen.

max-veit commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,76 @@
+necessary, add changes to the code to comply with any other added tests coming
+in from the updated master branch if they were not resolved in the merging or
+rebase process. Then you are ready to create your PR. Upon creating the PR our
+continuous integration spins up and does further checks with different compilers
```
Remove "spins up", it just runs a rigorous series of checks with different build parameters.

max-veit commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,76 @@
+continuous integration spins up and does further checks with different compilers
+and compiling modes such as debug.
+
+Now it is time for review. Github provides a button for review requests. Most of
```
Style: "Once your PR is ready for review (all items on the checklist are completed), use the Github button to notify specific developers to review the changes."

max-veit commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,76 @@
+This short description should serve as a rough guideline on how we deal with
+reviews on pull requests (PR). Some PRs are easy, others are hard. Some require
+multiple changes and continous feedback, others can be merged in flawlessly. So
+this text should communicate our intent and style and not be a go/no-go
```
Remove "So", or change to "Therefore". Also, "go/no-go" is unclear; perhaps "instead of being a strict set of rules". Also, instead of "should communicate", maybe "the purpose of this text is to communicate" is stronger.

max-veit commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,76 @@
+the time the suggested persons for review are a good first address. Add some so that they are notified.
+
+To help speed up the process we follow some rough guidelines so you know what
+you can expect from a reviewer's role and we know what we have to do:
```
Just "so everyone knows what to expect from a reviewer's role"

max-veit commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,76 @@
+ #. If someone starts a review, this person is the go-to person to finish the
+    review on the PR and should see it through as fast as possible. This means
+    following up the the requested changes and staying part of the
+    discussion. And it also includes the approval in the end. As fast as
```
Stronger: "Once the reviewers' concerns have been addressed, this person is responsible for approving them as soon as possible. This not only ensures that the changes are available to everyone as soon as they are ready, but also minimizes further merges with master and the associated extra work."

max-veit commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,76 @@
+    are conflicts with the PR, this of course has to be solved. Again, as a
+    courtesy, please ask for feedback on the changes again because
+    merges/rebases can be complicated and we want to avoid a broken master
+    branch.
```
Perhaps add "At the very least, make sure that all the CI checks pass before merging." (to avoid a borked master)

max-veit commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,76 @@
+    the process, make a proper review and take the responsibility to see it
+    through.
+
+Sometimes there is a lot of commit noise and the master changes and sometimes
```
I'm not sure what this means. Is this paragraph meant to warn about the difficulties that can arise _after_ review?

mastricker commented:
Thanks for the input. I think I have addressed all of the points.
I especially stressed the point to start a draft PR asap.

I also added a _TL;DR_ at the end, which sums up the points.

Let me know what you think.

max-veit commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,110 @@
+   please ask for feedback on the changes again because merges/rebases can be
+   complicated and we want to avoid a broken master branch.
+
+**TL;DR**
```
Let's merge this with the `CONTRIBUTING.md`, which can serve as a short reference (it's shown to potential contributors on opening their first pull request). That document should, in turn, link to here for the full description.

mastricker replied:
Yes, good idea. How do I make sure it is compiled into an html?
The links are hardcoded, right?

max-veit replied:
The `CONTRIBUTING` file can link to the online HTML documentation (since it will need to be followable from github, effectively an external site).

Not sure about the link in the other direction, since the CONTRIBUTING file isn't compiled as part of the doc build. You could have a look at `index.rst` which includes sections of the `README` for one way to do it (if you just want to automatically paste the summary points to the bottom of the `review_process` page). It would also be good to just reference the file by name near the beginning of the document.

mastricker commented:
I added all you proposed changes - though I am not sure about the linking of the TL;DR from `CONTRIBUTING.md` - is it correct how I did it?

Otherwise everything is in.

max-veit commented on CONTRIBUTING.rst:
```diff
@@ -28,6 +28,32 @@ request" instead (or waiting to request reviewers).
 .. [*] You'd think this would be obvious, but far too many people ignore
        ``README``\ s in practice
 
+The TL;DR of our review process
```
```suggestion
Summary of the review process
```
sounds a bit more professional.

max-veit commented on CONTRIBUTING.rst:
```diff
@@ -28,6 +28,32 @@ request" instead (or waiting to request reviewers).
+===============================
+
+A detailed description can be found in a separate document about the `review
+process <https://cosmo-epfl.github.io/librascal/dev_guide/review_process.html>`_
```
Ah, I didn't see this before. Yes, this looks like the right way to link _to_ the online HTML docs.

max-veit commented on CONTRIBUTING.rst:
```diff
@@ -28,6 +28,32 @@ request" instead (or waiting to request reviewers).
+ * Use our Pull-Request template in preparing your PR.
+ * Add tests for added functionality.
+ * Make sure your proposed changes pass the all existing tests, change (not
+   disable!) them if necessary.
```
s/disable/deactivate

max-veit commented on CONTRIBUTING.rst:
```diff
@@ -28,6 +28,32 @@ request" instead (or waiting to request reviewers).
+process <https://cosmo-epfl.github.io/librascal/dev_guide/review_process.html>`_
+
+For developers
+ * We want a clean and tested master branch that is why we review code and use
```
```suggestion
 * We want a clean and tested master branch, which is why we review code and use
```
(The alternative is to start a new sentence at "That". See also https://www.quickanddirtytips.com/education/grammar/which-versus-that-0)

max-veit commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,95 @@
+reviews on pull requests (PR). Some PRs are easy, others are hard. Some require
+multiple changes and continuous feedback, others can be merged in
+flawlessly. The purpose of this text is to communicate our intent and style and
+should not be seen as *law*.
```
Add: "A summary of these guidelines can be found in `CONTRUBTING.rst` in the top level of the repository. If you haven't yet read that file, it would be a good idea to do that now."

felixmusil commented on CONTRIBUTING.rst:
```diff
@@ -28,6 +28,32 @@ request" instead (or waiting to request reviewers).
+ * Start the discussion about your contribution early by using the
+   Draft-Pull-Request feature of Github.
+ * Use our Pull-Request template in preparing your PR.
+ * Add tests for added functionality.
```
```suggestion
 * Provide tests for the new/changed functionalities in the test suite of the library.
```

felixmusil commented on docs/source/dev_guide/how_to_contribute.rst:
```diff
@@ -38,4 +36,4 @@ At this point, you can code the method in the way that you want, but we remind y
 
-Once that the method that you need as been coded, you can save the file(s) and proceed. Note that adding a ``.cc`` file requires its registration in the folder's CMakeLists.txt so that it can be compiled.
\ No newline at end of file
+Once that the method that you need as been coded, you can save the file(s) and proceed. Note that adding a ``.cc`` file requires its registration in the folder's CMakeLists.txt so that it can be compiled.
```
```suggestion
Once that the method that you need as been implemented, you can save the file(s) and proceed. Note that adding a ``.cc`` file requires its registration in the folder's CMakeLists.txt for it to be compiled.
```

felixmusil commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,95 @@
+you should ensure that all existing and new tests pass. These include tests for
+functionality on the C++ as well as the python side but also include tests for
+code style as far as they can be automated (see our :ref:`Coding convention
+<coding-convention>` for details on the style). We also have a custom ``make``
```
```suggestion
<coding-convention>` for details on the style). We also have a custom ``make pretty-cpp`` and ``make pretty-python``
```

felixmusil commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,95 @@
+adding new documentation, in short: contributing to the codebase in any form,
+you should ensure that all existing and new tests pass. These include tests for
+functionality on the C++ as well as the python side but also include tests for
+code style as far as they can be automated (see our :ref:`Coding convention
```
```suggestion
code style as far as they can be automated for which the custom command is ``make lint`` (see our :ref:`Coding convention
```

felixmusil commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,95 @@
+discussion about the contribution and the proposed changes and it does not have
+to be polished at that stage. The difference between a Draft PR and a normal PR
+is minor. The Draft stage signals that your contribution is not ready for review
+but up for discussion. But it is nevertheless treated by like a normal PR in the
```
```suggestion
but up for discussion. But it is nevertheless treated like a normal PR in the
```

felixmusil commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,95 @@
+Once you have added something to the code base, please check locally if the
+tests pass. These tests include the code tests as well as the tests for code
+style as far as they can be checked automatically. You should also add tests for
+new functionality. Likewise, if your changes break existing tests, be sure to
```
```suggestion
new functionalities. Likewise, if your changes break existing tests, be sure to
```

max-veit replied:
Actually, the previous one was correct; "functionality" can't be pluralized in this context.

felixmusil commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,95 @@
+   should become the reviewer. This is to ensure that any part is reviewed by
+   someone else other than the creator.
+
+A last note related to our own experience. If it happens that your PR falls into
```
```suggestion
A last note related to our own experiences. If it happens that your PR falls into
```

max-veit replied:
Again, "experience" was correct.

mastricker replied:
Well this one did not change. I reverted the functionalities.

max-veit replied:
Ok cool, I didn't see that the suggestion hadn't actually been committed.

felixmusil commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,95 @@
+you, do not hesitate to contact us if the any issues arise *after* your PR has
+been approved. While we discourage changes after the PR has been approved,
+sometimes it is necessary to be able to merge with master. There are no rules to
+make this easy so please decide on the basis of the fact that we want a clean
```
```suggestion
make this easy so please make sure to keep a 'clean' and
```

max-veit replied:
Eh... I'd remove the quotes around "clean"

max-veit replied:
Same here (didn't see that the suggestion wasn't taken)

felixmusil commented on docs/source/dev_guide/review_process.rst:
```diff
@@ -0,0 +1,95 @@
+been approved. While we discourage changes after the PR has been approved,
+sometimes it is necessary to be able to merge with master. There are no rules to
+make this easy so please decide on the basis of the fact that we want a clean
+compiling master branch. To elaborate on this, here are a few examples
```
```suggestion
functional master branch. To elaborate on this, here are a few examples
```

mastricker commented:
All good. No comments other than parts of it also concerned already existing things.

max-veit commented on docs/source/dev_guide/how_to_contribute.rst:
```diff
@@ -38,4 +36,4 @@ At this point, you can code the method in the way that you want, but we remind y
 
-Once that the method that you need as been coded, you can save the file(s) and proceed. Note that adding a ``.cc`` file requires its registration in the folder's CMakeLists.txt so that it can be compiled.
\ No newline at end of file
+Once that the method that you need as been implemented, you can save the file(s) and proceed. Note that adding a ``.cc`` file requires its registration in the folder's CMakeLists.txt for it to be compiled.
```
```suggestion
Once the method that you need has been implemented, you can save the file(s) and proceed. Note that adding a ``.cc`` file requires its registration in the folder's CMakeLists.txt for it to be compiled.
```

max-veit commented on CONTRIBUTING.rst:
```diff
@@ -28,6 +28,32 @@ request" instead (or waiting to request reviewers).
+ * Start the discussion about your contribution early by using the
+   Draft-Pull-Request feature of Github.
+ * Use our Pull-Request template in preparing your PR.
+ * Provide tests for the new/changed functionalities in the test suite of the library.
```
```suggestion
 * Provide tests for the new/changed functionality in the test suite of the library.
```
"functionality" is a mass noun in this context and therefore can't be pluralized. Plus, "functionalities" just sounds kind of weird.

mastricker commented:
Reverted and changed last commit, I'll wait for the CI to go through...
